packages/peregrine/lib/talons/SignIn/__tests__/useSignIn.spec.js

```diff
@@ -0,0 +1,174 @@
+jest.mock('../../../hooks/useAwaitQuery');
+jest.mock('../../../store/actions/cart', () => ({
+    retrieveCartId: jest.fn().mockReturnValue('new-cart-id')
+}));
+
+jest.mock('../../../context/cart', () => ({
+    useCartContext: jest.fn().mockReturnValue([
+        { cartId: 'old-cart-id' },
+        {
+            createCart: jest.fn(),
+            removeCart: jest.fn(),
+            getCartDetails: jest.fn()
+        }
+    ])
+}));
+
+jest.mock('../../../context/user', () => ({
+    useUserContext: jest.fn().mockReturnValue([
+        {
+            isGettingDetails: false,
+            getDetailsError: 'getDetails error from redux'
+        },
+        { getUserDetails: jest.fn(), setToken: jest.fn() }
+    ])
+}));
+
+const signInVariables = {
+    email: 'fry@planetexpress.com',
+    password: 'slurm is the best'
+};
+const authToken = 'auth-token-123';
+
+const signInMock = {
+    request: {
+        query: defaultOperations.signInMutation,
+        variables: signInVariables
+    },
+    result: {
+        data: { generateCustomerToken: { token: authToken } }
+    }
+};
+
+const mergeCartsMock = {
+    request: {
+        query: defaultOperations.mergeCartsMutation,
+        variables: {
+            destinationCartId: 'new-cart-id',
+            sourceCartId: 'old-cart-id'
+        }
+    },
+    result: {
+        data: null
+    }
+};
+
+const initialProps = {
+    getCartDetailsQuery: 'getCartDetailsQuery',
+    setDefaultUsername: jest.fn(),
+    showCreateAccount: jest.fn(),
+    showForgotPassword: jest.fn()
+};
+
+const renderHookWithProviders = ({
+    renderHookOptions = { initialProps },
+    mocks = [signInMock, mergeCartsMock]
+} = {}) => {
+    const wrapper = ({ children }) => (
+        <MockedProvider mocks={mocks} addTypename={false}>
+            {children}
+        </MockedProvider>
+    );
+
+    return renderHook(useSignIn, { wrapper, ...renderHookOptions });
+};
+
+test('returns correct shape', () => {
+    const { result } = renderHookWithProviders();
+
+    expect(result.current).toMatchInlineSnapshot(`
+        Object {
+          "errors": Map {
+            "getUserDetailsQuery" => "getDetails error from redux",
+            "signInMutation" => undefined,
+          },
+          "handleCreateAccount": [Function],
+          "handleForgotPassword": [Function],
+          "handleSubmit": [Function],
+          "isBusy": false,
+          "setFormApi": [Function],
+        }
+    `);
+});
+
+test('handleSubmit triggers waterfall of operations and actions', async () => {
+    const [, { getCartDetails }] = useCartContext();
+    const [, { getUserDetails, setToken }] = useUserContext();
+
+    const { result } = renderHookWithProviders();
+
+    await act(() => result.current.handleSubmit(signInVariables));
+
+    expect(result.current.isBusy).toBe(true);
+    expect(setToken).toHaveBeenCalledWith(authToken);
+    expect(getCartDetails).toHaveBeenCalled();
+    expect(getUserDetails).toHaveBeenCalled();
```

Comment:
I would like to avoid implementation details from bleeding into tests, but included these two checks since we're mocking both of these contexts. Ideally we wouldn't mock these contexts and include them in the helper `renderHookWithProviders`, but that proved to be a bit difficult with redux in place. I hope once redux is deprecated from these contexts, this will be easier going forward, and we can assert on global state results, instead of checking that methods were called.